aws-lc-rs/src/kem.rs

```diff
@@ -0,0 +1,575 @@
+    ///
+    /// `bytes` is a slice of raw bytes representing a `KemPrivateKey`
+    ///
+    /// # Errors
+    /// `error::KeyRejected` when operation fails during key creation.
+    pub fn new(alg: &'static Algorithm, bytes: &[u8]) -> Result<Self, KeyRejected> {
+        match bytes.len().cmp(&alg.secret_key_size()) {
+            Ordering::Less => Err(KeyRejected::too_small()),
+            Ordering::Greater => Err(KeyRejected::too_large()),
+            Ordering::Equal => Ok(()),
+        }?;
+        let privkey = LcPtr::new(unsafe {
+            EVP_PKEY_kem_new_raw_secret_key(alg.id.nid(), bytes.as_ptr(), bytes.len())
+        })?;
+        Ok(PrivateKey {
+            algorithm: alg,
+            evp_pkey: privkey,
+            priv_key: bytes.into(),
+        })
+    }
+}
+
+impl Drop for PrivateKey {
+    fn drop(&mut self) {
+        self.priv_key.zeroize();
+    }
+}
+
+impl AsRef<[u8]> for PrivateKey {
+    fn as_ref(&self) -> &[u8] {
+        &self.priv_key
+    }
+}
+
+/// A serializable public key usable with KEMS. This can be constructed
+/// from a `KemPrivateKey` or constructed from raw bytes.
+#[derive(Debug)]
+pub struct PublicKey {
+    algorithm: &'static Algorithm,
+    evp_pkey: LcPtr<EVP_PKEY>,
+    pub_key: Box<[u8]>,
+}
+
+impl PublicKey {
+    /// Return the algorithm associated with the given KEM public key.
+    #[must_use]
+    pub fn algorithm(&self) -> &'static Algorithm {
+        self.algorithm
+    }
+
+    /// Performs the encapsulate operation using the current KEM public key, generating a ciphertext
+    /// and associated shared secret.
+    ///
+    /// After the encapsulation is finished, `encapsulate` calls `kdf` with the ciphertext and raw shared secret
+    /// from the operation and then returns what `kdf` returns.
+    ///
+    /// # Errors
+    /// `error::Unspecified` when operation fails due to internal error.
+    #[allow(clippy::missing_panics_doc)]
+    pub fn encapsulate<F, R, E>(&self, error_value: E, kdf: F) -> Result<R, E>
+    where
+        F: FnOnce(&[u8], &[u8]) -> Result<R, E>,
+    {
+        let mut ciphertext_len = self.algorithm.ciphertext_size();
+        let mut shared_secret_len = self.algorithm.shared_secret_size();
+        let mut ciphertext: Vec<u8> = vec![0u8; ciphertext_len];
+        let mut shared_secret: Vec<u8> = vec![0u8; shared_secret_len];
+
+        let ctx =
+            if let Ok(ctx) = LcPtr::new(unsafe { EVP_PKEY_CTX_new(*self.evp_pkey, null_mut()) }) {
+                ctx
+            } else {
+                return Err(error_value);
+            };
+
+        if 1 != unsafe {
+            EVP_PKEY_encapsulate(
+                *ctx,
+                ciphertext.as_mut_ptr(),
+                &mut ciphertext_len,
+                shared_secret.as_mut_ptr(),
+                &mut shared_secret_len,
+            )
+        } {
+            return Err(error_value);
+        }
+
+        kdf(&ciphertext, &shared_secret)
```

Comment:
so in `decaps` you call `kdf(shared_secret)` and here you call `kdf(ciphertext, shared_secret)`. But the output of these two function calls should be the same for the protocol to work. How does this even pass the tests?

Reply:
Addressed in the earlier comment about the `kdf` argument.